src/alloc.rs

```diff
@@ -382,6 +372,58 @@ fn exit_alloc(msg: &str) {
     std::process::exit(1);
 }
 
+// -----------------------------------------------------------------------------
+//        APIs for querying the block information about each allocation
+// -----------------------------------------------------------------------------
+
+pub struct AllocMetadata;
+
+impl AllocMetadata {
+    /// Insert metadata about an allocation block. The GC scans for reachable
+    /// objects conservatively. It uses the metadata stored against each
+    /// allocation to determine:
+    /// 1) the size of an allocation block
+    /// 2) whether an allocation block's memory management is the responsibility
+    ///    of the GC
+    /// 3) whether an arbitrary pointer-like word points within an
+    ///    allocation block
+    pub(crate) fn insert(ptr: usize, size: usize, gc: bool) {
+        ALLOC_LOCK.lock();
+        unsafe {
+            match ALLOC_INFO {
+                Some(ref mut pm) => pm.insert(ptr as usize, size, true),
+                None => {
+                    let mut al = AllocList::new();
+                    al.insert(ptr as usize, size, gc);
+                    ALLOC_INFO = Some(al);
+                }
+            };
+        }
+        ALLOC_LOCK.unlock();
+    }
+    /// Returns metadata about an allocation block when given an arbitrary
+    /// pointer inside that block. Either the original pointer from the call to
+    /// `alloc` can be passed, or an inner pointer to some offset within the
+    /// block.
+    pub(crate) fn find(ptr: usize) -> Option<PtrInfo> {
+        ALLOC_LOCK.lock();
+        let ptr_info = unsafe { ALLOC_INFO.as_ref().unwrap().find_base(ptr) };
+        ALLOC_LOCK.unlock();
+        ptr_info
+    }
+
+    /// Removes the metadata associated with an allocation.
+    pub(crate) fn remove(ptr: usize) {
+        ALLOC_LOCK.lock();
+        let ptr_info = unsafe { ALLOC_INFO.as_ref().unwrap().find_base(ptr).unwrap() };
+
+        debug_assert!(ptr_info.gc); // FIXME: Move this expensive function into the assertion
```

Comment:
I don't understand the assertion: the `assert` isn't a function call. You can, from memory, do things like:

```rust
#[cfg(debug)]
{
    let ptr_info = ...;
    assert!(ptr_info.gc);
}
```

if that's what you wanted?

Reply:
Sorry, that comment shouldn't have ended up in the PR and was a reminder for me to do exactly what you suggested.